File: android/src/main/kotlin/com/monorithm/monolens/VideoExporter.kt
package com.monorithm.monolens

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.annotation.OptIn
import androidx.media3.common.Effect
import androidx.media3.common.MediaItem
import androidx.media3.common.util.UnstableApi
import androidx.media3.effect.BitmapOverlay
import androidx.media3.effect.Crop
import androidx.media3.effect.OverlayEffect
import androidx.media3.effect.ScaleAndRotateTransformation
import androidx.media3.effect.TextureOverlay
import androidx.media3.transformer.Composition
import androidx.media3.transformer.Effects
import androidx.media3.transformer.EditedMediaItem
import androidx.media3.transformer.ExportException
import androidx.media3.transformer.ExportResult
import androidx.media3.transformer.ProgressHolder
import androidx.media3.transformer.Transformer
import com.google.common.collect.ImmutableList
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Trims clips with Media3 Transformer and pulls filmstrip frames.
 *
 * Transformer is the ffmpeg-free path: it drives the platform's own MediaCodec
 * encoders, so there is no bundled native binary, no per-ABI APK cost, and the
 * work lands on hardware. `ffmpeg_kit` being retired is what makes this the
 * only maintained option on Android, and it is the better one anyway.
 *
 * Every Transformer call goes through [mainHandler]: the class is single-
 * threaded on the looper it was built on, and cancelling from another thread
 * throws. [jobs] is therefore main-thread-confined and needs no lock.
 */
@OptIn(UnstableApi::class)
internal class VideoExporter(
    private val context: Context,
    private val onProgress: (String, Double) -> Unit,
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    private val jobs = mutableMapOf<String, Job>()

    /** Cancels that arrived before their export started. */
    private val cancelledBeforeStart = mutableSetOf<String>()

    private class Job(
        val callback: (Result<MediaInfo>) -> Unit,
    ) {
        var transformer: Transformer? = null
        var ticker: Runnable? = null
        var settled = false
    }

    fun trim(
        request: VideoTrimRequest,
        jobId: String,
        callback: (Result<MediaInfo>) -> Unit,
    ) {
        mainHandler.post {
            val job = Job(callback)
            jobs[jobId] = job
            try {
                start(request, jobId, job)
            } catch (error: MonolensException) {
                settle(jobId, Result.failure(error))
            } catch (error: Exception) {
                settle(
                    jobId,
                    Result.failure(
                        MonolensException.exportFailed(error.message ?: "Export failed"),
                    ),
                )
            }
        }
    }

    private fun start(request: VideoTrimRequest, jobId: String, job: Job) {
        val source = File(request.sourcePath)
        if (!source.exists()) throw MonolensException.notFound(request.sourcePath)

        if (cancelledBeforeStart.remove(jobId)) {
            // cancel() beat us here — honour it rather than starting the work.
            throw MonolensException.cancelled()
        }

        val durationMs = MediaProbe.probe(request.sourcePath).durationMs ?: 0L
        val startMs = max(0L, request.startMs)
        val endMs = if (durationMs > 0) minOf(request.endMs, durationMs) else request.endMs
        if (endMs <= startMs) {
            throw MonolensException.invalidRange(
                "Range ${request.startMs}-${request.endMs} ms is empty for a $durationMs ms clip",
            )
        }

        val output = File(request.outputPath)
        output.parentFile?.mkdirs()
        if (output.exists()) output.delete()

        val mediaItem = MediaItem.Builder()
            .setUri(Uri.fromFile(source))
            .setClippingConfiguration(
                MediaItem.ClippingConfiguration.Builder()
                    .setStartPositionMs(startMs)
                    .setEndPositionMs(endMs)
                    // Leaving this false forces a re-encode so the cut lands on
                    // the requested frame. Snapping to a sync sample would slide
                    // the in-point backwards by up to a GOP, which is wrong when
                    // the author has just placed a handle by eye.
                    .setStartsAtKeyFrame(false)
                    .build(),
            )
            .build()

        val editedMediaItem = EditedMediaItem.Builder(mediaItem)
            // Dropping the track, not exporting silence — a silent AAC track is
            // bytes the upload does not need.
            .setRemoveAudio(request.muteAudio)
            .setEffects(Effects(emptyList(), videoEffects(request)))
            .build()

        val transformer = Transformer.Builder(context)
            .addListener(object : Transformer.Listener {
                override fun onCompleted(composition: Composition, exportResult: ExportResult) {
                    onProgress(jobId, 1.0)
                    settle(
                        jobId,
                        runCatching {
                            // Report the range that was asked for: the container's
                            // own duration can land a frame either side after a
                            // re-encode.
                            MediaProbe.probe(request.outputPath)
                                .copy(durationMs = endMs - startMs)
                        },
                    )
                }

                override fun onError(
                    composition: Composition,
                    exportResult: ExportResult,
                    exportException: ExportException,
                ) {
                    settle(
                        jobId,
                        Result.failure(
                            MonolensException.exportFailed(describe(exportException)),
                        ),
                    )
                }
            })
            .build()

        val ticker = object : Runnable {
            private val holder = ProgressHolder()

            override fun run() {
                val current = jobs[jobId]?.transformer ?: return
                if (current.getProgress(holder) == Transformer.PROGRESS_STATE_AVAILABLE) {
                    onProgress(jobId, holder.progress / 100.0)
                }
                mainHandler.postDelayed(this, PROGRESS_INTERVAL_MS)
            }
        }

        job.transformer = transformer
        job.ticker = ticker
        transformer.start(editedMediaItem, request.outputPath)
        mainHandler.postDelayed(ticker, PROGRESS_INTERVAL_MS)
    }

    /**
     * Flattens an exception's cause chain into the message.
     *
     * Media3 reports the top of the chain, which for anything in the frame
     * pipeline is the useless "Video frame processing error" -- the actual
     * reason, such as a shader that failed to compile, is two causes down.
     */
    private fun describe(error: Throwable): String {
        val parts = mutableListOf<String>()
        var current: Throwable? = error
        var depth = 0
        while (current != null && depth < 5) {
            current.message?.takeIf { it.isNotBlank() }?.let(parts::add)
            current = current.cause
            depth++
        }
        return parts.distinct().joinToString(": ").ifBlank { "Export failed" }
    }

    /**
     * The frame pipeline: crop, then rotate, then the flattened overlay.
     *
     * Order matters and mirrors the still path exactly, because an annotation
     * placed against the output frame has to land in the same place on both.
     * Media3 applies effects in list order, so this reads as it runs.
     */
    private fun videoEffects(request: VideoTrimRequest): List<Effect> {
        val effects = mutableListOf<Effect>()

        request.crop?.let { rect ->
            // Crop takes normalized device coordinates: -1..1, origin centre,
            // y up. The wire rect is 0..1 from the top-left.
            effects += Crop(
                (2 * rect.left - 1).toFloat(),
                (2 * (rect.left + rect.width) - 1).toFloat(),
                (1 - 2 * (rect.top + rect.height)).toFloat(),
                (1 - 2 * rect.top).toFloat(),
            )
        }

        val degrees = when (request.rotation) {
            MonoRotation.NONE -> 0f
            MonoRotation.QUARTER_TURN -> 90f
            MonoRotation.HALF_TURN -> 180f
            MonoRotation.THREE_QUARTER_TURNS -> 270f
        }
        if (degrees != 0f) {
            // Media3 rotates counter-clockwise for a positive angle; MonoRotation
            // names a clockwise turn.
            effects += ScaleAndRotateTransformation.Builder()
                .setRotationDegrees(-degrees)
                .build()
        }

        // Blur before the overlay and after the geometry: it samples the frame,
        // so it belongs to the picture, and its regions are normalized against
        // the cropped and rotated output. A caption drawn over a blurred face
        // then stays legible, matching the still path.
        val blurred = request.annotations.filter { it.kind == AnnotationKind.BLUR }
        if (blurred.isNotEmpty()) {
            effects += MaskedBlurEffect(blurred)
        }

        if (request.flipHorizontal) {
            // A separate effect rather than setScale on the rotation above:
            // Media3 applies a list in order, so this mirrors the already
            // rotated frame. Folding both into one transformation would scale
            // first and mirror the *source*, which is the opposite convention
            // to the still path.
            effects += ScaleAndRotateTransformation.Builder()
                .setScale(-1f, 1f)
                .build()
        }

        val painted = request.annotations.filter { it.kind != AnnotationKind.BLUR }
        if (painted.isNotEmpty()) {
            val size = outputSize(request)
            AnnotationRenderer.overlayLayer(request.annotations, size.first, size.second)
                ?.let { layer ->
                    effects += OverlayEffect(
                        ImmutableList.of<TextureOverlay>(
                            BitmapOverlay.createStaticBitmapOverlay(layer),
                        ),
                    )
                }
        }

        return effects
    }

    /**
     * The frame size the overlay has to be rasterized at — the source's display
     * size with the crop and the rotation applied, which is what the annotation
     * coordinates were normalized against.
     */
    private fun outputSize(request: VideoTrimRequest): Pair<Int, Int> {
        val info = runCatching { MediaProbe.probe(request.sourcePath) }.getOrNull()
        var width = (info?.width ?: 1080L).toInt()
        var height = (info?.height ?: 1920L).toInt()

        request.crop?.let { rect ->
            width = max(1, (width * rect.width).roundToInt())
            height = max(1, (height * rect.height).roundToInt())
        }
        val turned = request.rotation == MonoRotation.QUARTER_TURN ||
            request.rotation == MonoRotation.THREE_QUARTER_TURNS
        return if (turned) height to width else width to height
    }

    /**
     * `Transformer.cancel()` fires no listener callback, so the Dart future
     * waiting on this job would hang forever if cancel did not settle it here.
     */
    fun cancel(jobId: String) {
        mainHandler.post {
            val job = jobs[jobId]
            if (job == null) {
                // The export has not started yet; remember so start() refuses to.
                cancelledBeforeStart.add(jobId)
                return@post
            }
            runCatching { job.transformer?.cancel() }
            settle(jobId, Result.failure(MonolensException.cancelled()))
        }
    }

    /** Completes a job exactly once — a second completion would throw in Dart. */
    private fun settle(jobId: String, result: Result<MediaInfo>) {
        val job = jobs[jobId] ?: return
        if (job.settled) return
        job.settled = true
        jobs.remove(jobId)
        job.ticker?.let(mainHandler::removeCallbacks)
        job.callback(result)
    }

    // Filmstrip

    fun thumbnails(path: String, atMs: List<Long>, maxDimension: Long): List<ByteArray> {
        if (!File(path).exists()) throw MonolensException.notFound(path)

        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(path)
            val size = maxDimension.toInt().coerceAtLeast(1)

            atMs.map { milliseconds ->
                // OPTION_CLOSEST_SYNC decodes from the nearest keyframe rather
                // than walking forward to an exact frame. A filmstrip is a
                // scrubbing aid, and exact seeking here is an order of magnitude
                // slower for no visible gain.
                val timeUs = max(0L, milliseconds) * 1000
                val frame = runCatching {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                        // Decodes straight to the thumbnail size. The fallback
                        // below decodes the frame at full resolution and then
                        // throws most of it away.
                        retriever.getScaledFrameAtTime(
                            timeUs,
                            MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
                            size,
                            size,
                        )
                    } else {
                        retriever.getFrameAtTime(
                            timeUs,
                            MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
                        )
                    }
                }.getOrNull()

                // One unreadable frame should not fail the strip; an empty entry
                // is the caller's cue to draw a placeholder.
                frame?.let { encode(it, size) } ?: ByteArray(0)
            }
        } finally {
            runCatching { retriever.release() }
        }
    }

    private fun encode(frame: Bitmap, maxDimension: Int): ByteArray {
        val longest = max(frame.width, frame.height)
        val scaled = if (maxDimension in 1 until longest) {
            val factor = maxDimension.toFloat() / longest
            Bitmap.createScaledBitmap(
                frame,
                (frame.width * factor).toInt().coerceAtLeast(1),
                (frame.height * factor).toInt().coerceAtLeast(1),
                true,
            )
        } else {
            frame
        }

        return ByteArrayOutputStream().use { stream ->
            scaled.compress(Bitmap.CompressFormat.JPEG, 70, stream)
            if (scaled != frame) scaled.recycle()
            frame.recycle()
            stream.toByteArray()
        }
    }

    private companion object {
        const val PROGRESS_INTERVAL_MS = 100L
    }
}

File: ios/monolens/Sources/monolens/VideoExporter.swift
import AVFoundation
import UIKit

/// Trims clips with AVFoundation and pulls filmstrip frames.
///
/// Deliberately not passthrough: `AVAssetExportPresetPassthrough` only cuts on
/// sync samples, so an in-point between keyframes silently slides backwards by
/// up to a GOP — fine for a rough cut, wrong when the author has just placed a
/// handle on a specific frame. Re-encoding costs a second or two on a short
/// clip and puts the cut where they asked for it.
final class VideoExporter {

  /// In-flight exports, so `cancel(jobId:)` can reach one. Touched from the
  /// platform thread and from export callbacks — hence the queue.
  private var sessions: [String: AVAssetExportSession] = [:]
  private var cancelledJobs: Set<String> = []
  private let lock = DispatchQueue(label: "com.monorithm.monolens.exports")

  private let onProgress: (String, Double) -> Void

  init(onProgress: @escaping (String, Double) -> Void) {
    self.onProgress = onProgress
  }

  func trim(
    request: VideoTrimRequest,
    jobId: String,
    completion: @escaping (Result<MediaInfo, Error>) -> Void
  ) {
    let sourceURL = URL(fileURLWithPath: request.sourcePath)
    guard FileManager.default.fileExists(atPath: request.sourcePath) else {
      completion(.failure(MonolensError.notFound(request.sourcePath)))
      return
    }

    let asset = AVURLAsset(url: sourceURL)
    let durationMs = Int64((CMTimeGetSeconds(asset.duration) * 1000).rounded())
    let startMs = max(0, request.startMs)
    let endMs = min(request.endMs, durationMs)
    guard endMs > startMs else {
      completion(
        .failure(
          MonolensError.invalidRange(
            "Range \(request.startMs)-\(request.endMs) ms is empty for a \(durationMs) ms clip")))
      return
    }

    // Milliseconds against a 1000-tick timescale: exact for the integer values
    // the Dart side works in, so no drift creeps into the cut point.
    let start = CMTime(value: CMTimeValue(startMs), timescale: 1000)
    let end = CMTime(value: CMTimeValue(endMs), timescale: 1000)
    let range = CMTimeRange(start: start, end: end)

    let exportAsset: AVAsset
    let exportRange: CMTimeRange
    if request.muteAudio {
      // Muting means dropping the track, not exporting silence — a silent AAC
      // track is bytes the upload does not need.
      do {
        exportAsset = try videoOnlyComposition(from: asset, range: range)
        exportRange = CMTimeRange(start: .zero, duration: range.duration)
      } catch {
        completion(.failure(error))
        return
      }
    } else {
      exportAsset = asset
      exportRange = range
    }

    guard
      let session = AVAssetExportSession(
        asset: exportAsset, presetName: AVAssetExportPresetHighestQuality)
    else {
      completion(.failure(MonolensError.exportFailed("No exporter for this asset")))
      return
    }

    let outputURL = URL(fileURLWithPath: request.outputPath)
    try? FileManager.default.removeItem(at: outputURL)

    session.outputURL = outputURL
    session.outputFileType =
      session.supportedFileTypes.contains(.mp4) ? .mp4 : .mov
    session.timeRange = exportRange

    // Only build a composition when the frame actually changes. A plain trim
    // stays on the straight path, which is markedly faster.
    if request.crop != nil || request.rotation != .none || request.flipHorizontal
      || !request.annotations.isEmpty
    {
      session.videoComposition = Self.videoComposition(for: exportAsset, request: request)
    }
    // Lets the OS trim to the requested range rather than shipping the whole
    // source into the pipeline.
    session.shouldOptimizeForNetworkUse = true

    var alreadyCancelled = false
    lock.sync {
      if cancelledJobs.contains(jobId) {
        // cancel() beat us here — honour it rather than starting the work.
        cancelledJobs.remove(jobId)
        alreadyCancelled = true
      } else {
        sessions[jobId] = session
      }
    }
    if alreadyCancelled {
      completion(.failure(MonolensError.cancelled))
      return
    }

    let ticker = startTicking(jobId: jobId, session: session)

    session.exportAsynchronously { [weak self] in
      guard let self else { return }
      ticker.cancel()
      self.lock.sync { _ = self.sessions.removeValue(forKey: jobId) }

      switch session.status {
      case .completed:
        self.onProgress(jobId, 1.0)
        do {
          var info = try MediaProbe.probe(path: request.outputPath)
          // Report the range that was asked for: the container's own duration
          // can land a frame either side after a re-encode.
          info.durationMs = endMs - startMs
          completion(.success(info))
        } catch {
          completion(.failure(error))
        }
      case .cancelled:
        completion(.failure(MonolensError.cancelled))
      default:
        let message = session.error?.localizedDescription ?? "Export failed"
        completion(.failure(MonolensError.exportFailed(message)))
      }
    }
  }

  func cancel(jobId: String) {
    lock.sync {
      if let session = sessions[jobId] {
        session.cancelExport()
      } else {
        // The export has not started yet; remember so trim() refuses to start.
        cancelledJobs.insert(jobId)
      }
    }
  }

  /// Polls `progress` on a timer. The property is a plain float with no KVO
  /// contract worth relying on, so a 10 Hz poll is the honest way to drive a
  /// progress bar.
  private func startTicking(jobId: String, session: AVAssetExportSession)
    -> DispatchSourceTimer
  {
    let timer = DispatchSource.makeTimerSource(queue: lock)
    timer.schedule(deadline: .now() + 0.1, repeating: 0.1)
    timer.setEventHandler { [weak self, weak session] in
      guard let self, let session else { return }
      self.onProgress(jobId, Double(session.progress))
    }
    timer.resume()
    return timer
  }

  /// Crop, rotate, blur and overlay, as one Core Image pass per frame.
  ///
  /// Core Image rather than layer instructions plus `AVVideoCompositionCoreAnimationTool`:
  /// blur has to sample the frame, which a CALayer overlay cannot do, and doing
  /// half the work in one mechanism and half in the other means two coordinate
  /// systems to keep in agreement. One pipeline, one set of rules.
  private static func videoComposition(
    for asset: AVAsset, request: VideoTrimRequest
  ) -> AVVideoComposition? {
    guard let track = asset.tracks(withMediaType: .video).first else { return nil }

    let natural = track.naturalSize
    let transformed = natural.applying(track.preferredTransform)
    let display = CGSize(width: abs(transformed.width), height: abs(transformed.height))
    guard display.width >= 1, display.height >= 1 else { return nil }

    // Crop in display space. Core Image is y-up and the normalized rect is
    // measured from the top, so the origin flips.
    let crop: CGRect
    if let rect = request.crop {
      crop = CGRect(
        x: CGFloat(rect.left) * display.width,
        y: CGFloat(1 - rect.top - rect.height) * display.height,
        width: max(1, CGFloat(rect.width) * display.width),
        height: max(1, CGFloat(rect.height) * display.height))
    } else {
      crop = CGRect(origin: .zero, size: display)
    }

    let turned = request.rotation == .quarterTurn || request.rotation == .threeQuarterTurns
    let output = turned
      ? CGSize(width: crop.height, height: crop.width)
      : crop.size
    // Encoders reject odd dimensions on some profiles; round to even.
    let renderSize = CGSize(
      width: max(2, (output.width / 2).rounded() * 2),
      height: max(2, (output.height / 2).rounded() * 2))

    let mask = AnnotationRenderer.blurMask(request.annotations, size: renderSize)
    let radius = AnnotationRenderer.blurRadius(request.annotations, size: renderSize)
    let overlay = AnnotationRenderer.overlayLayer(request.annotations, size: renderSize)
    let rotation = request.rotation
    let flipHorizontal = request.flipHorizontal
    let preferred = track.preferredTransform

    let composition = AVVideoComposition(asset: asset) { requestFrame in
      var image = requestFrame.sourceImage

      // Whether the source arrives already oriented is version-dependent, so
      // measure rather than assume: if the extent still matches the encoded
      // size while the display size differs, the transform is ours to apply.
      if abs(image.extent.width - natural.width) < 1,
        abs(image.extent.height - natural.height) < 1,
        display != natural
      {
        image = image.transformed(by: preferred)
      }
      image = Self.movedToOrigin(image)

      image = image.cropped(to: crop)
      image = Self.movedToOrigin(image)

      image = Self.rotated(image, by: rotation)
      image = Self.movedToOrigin(image)

      // After the rotation, so the mirror is about the *output's* vertical
      // axis — the same order the still path composes in, which is what makes
      // a rotate-and-flip land identically on a JPEG and on every frame here.
      if flipHorizontal {
        image = image.transformed(by: CGAffineTransform(scaleX: -1, y: 1))
        image = Self.movedToOrigin(image)
      }

      // renderSize was rounded to even for the encoder, so nudge the frame onto
      // it exactly — the mask and the overlay were rasterized at that size and
      // a one-pixel disagreement shows up as a seam.
      let scaleX = renderSize.width / max(1, image.extent.width)
      let scaleY = renderSize.height / max(1, image.extent.height)
      if abs(scaleX - 1) > 0.0001 || abs(scaleY - 1) > 0.0001 {
        image = image.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))
        image = Self.movedToOrigin(image)
      }

      image = AnnotationRenderer.applyingBlur(to: image, mask: mask, radius: radius)
      image = AnnotationRenderer.compositing(overlay, over: image)

      requestFrame.finish(with: image, context: nil)
    }

    // The CI-filter initializer hands back an immutable composition, and
    // renderSize is the one thing we must set.
    guard let mutable = composition.mutableCopy() as? AVMutableVideoComposition else {
      return composition
    }
    mutable.renderSize = renderSize
    return mutable
  }

  private static func movedToOrigin(_ image: CIImage) -> CIImage {
    image.transformed(
      by: CGAffineTransform(
        translationX: -image.extent.origin.x, y: -image.extent.origin.y))
  }

  /// Core Image is y-up, so a clockwise turn on screen is a negative angle.
  private static func rotated(_ image: CIImage, by rotation: MonoRotation) -> CIImage {
    let radians: CGFloat
    switch rotation {
    case .none: return image
    case .quarterTurn: radians = -.pi / 2
    case .halfTurn: radians = .pi
    case .threeQuarterTurns: radians = .pi / 2
    }
    return image.transformed(by: CGAffineTransform(rotationAngle: radians))
  }

  private func videoOnlyComposition(from asset: AVAsset, range: CMTimeRange) throws
    -> AVComposition
  {
    let composition = AVMutableComposition()
    guard let sourceTrack = asset.tracks(withMediaType: .video).first,
      let track = composition.addMutableTrack(
        withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid)
    else {
      throw MonolensError.exportFailed("No video track to export")
    }
    try track.insertTimeRange(range, of: sourceTrack, at: .zero)
    // Carry the rotation across, or a portrait clip exports on its side.
    track.preferredTransform = sourceTrack.preferredTransform
    return composition
  }

  // MARK: - Filmstrip

  /// Decodes every requested frame in one batch.
  ///
  /// `generateCGImagesAsynchronously` hands the whole time list to AVFoundation
  /// at once, so it walks the file in order and reuses one decode session;
  /// `copyCGImage` per frame re-seeks from scratch each time, which on a long
  /// clip is most of the cost of drawing a filmstrip.
  func thumbnails(path: String, atMs: [Int64], maxDimension: Int64) throws -> [Data] {
    let asset = AVURLAsset(url: URL(fileURLWithPath: path))
    guard asset.tracks(withMediaType: .video).first != nil else {
      throw MonolensError.unsupported(path)
    }

    let generator = AVAssetImageGenerator(asset: asset)
    generator.appliesPreferredTrackTransform = true
    generator.maximumSize = CGSize(width: CGFloat(maxDimension), height: CGFloat(maxDimension))
    // A filmstrip is a scrubbing aid — half a second of slop buys a large speed
    // win over frame-exact seeking, which would decode from the last keyframe
    // for every thumbnail.
    generator.requestedTimeToleranceBefore = CMTime(value: 1, timescale: 2)
    generator.requestedTimeToleranceAfter = CMTime(value: 1, timescale: 2)

    // Callbacks arrive out of order, keyed only by the time that was asked for,
    // so keep a queue of destination slots per timestamp — duplicated
    // timestamps then still land somewhere sensible.
    var slots: [CMTimeValue: [Int]] = [:]
    var times: [NSValue] = []
    for (index, milliseconds) in atMs.enumerated() {
      let value = CMTimeValue(max(0, milliseconds))
      slots[value, default: []].append(index)
      times.append(NSValue(time: CMTime(value: value, timescale: 1000)))
    }

    // One unreadable frame should not fail the strip; an empty entry is the
    // caller's cue to draw a placeholder.
    var frames = [Data](repeating: Data(), count: atMs.count)
    let lock = NSLock()
    let group = DispatchGroup()
    times.forEach { _ in group.enter() }

    generator.generateCGImagesAsynchronously(forTimes: times) {
      requested, image, _, _, _ in
      defer { group.leave() }
      guard let image,
        let data = UIImage(cgImage: image).jpegData(compressionQuality: 0.7)
      else { return }

      lock.lock()
      defer { lock.unlock() }
      if let index = slots[requested.value]?.first {
        slots[requested.value]?.removeFirst()
        frames[index] = data
      }
    }

    group.wait()
    return frames
  }
}
